File: 3803-count-residue-prefixes/3803-count-residue-prefixes.cs
public class Solution {
    public int ResiduePrefixes(string s) {
        var n = s.Length;
        int dis = 0, ans = 0;
        var freq = new int[26];

        for (var i = 0; i < n; ++i) {
            int ch = s[i] - 97;
            if (freq[ch] == 0)
                dis++;
            freq[ch]++;

            ans += ((dis == (i + 1) % 3) ? 1 : 0); 
        }
        return ans;
    }
}

File: 3803-count-residue-prefixes/3803-count-residue-prefixes.ts
/**
* Counts prefixes where:
*   number of distinct elements === (prefix length mod 3)
*/
function residuePrefixes<T>(sequence: Iterable<T>): number {
    const freq = new Map<T, number>();
    let distinct = 0;
    let index = 0;
    let answer = 0;

    for (const item of sequence) {
        index++;

        const count = freq.get(item) ?? 0;
        if (count === 0) {
            distinct++;
        }
        freq.set(item, count + 1);

        if (distinct === index % 3) {
            answer++;
        }
    }

    return answer;
};